Sort profile refresh tokens by 'last used at' date (#4484)

aav7fl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

The user profile page lists are tokens created by/used by said user. It's a bit of a mess when trying to find tokens that were recently used (and clearing out older tokens). This change sorts in ascending order the tokens by their `last used at` date.

Previously there was no sorting mechanism (although it was sorted by chance with `created dated` it appears).

## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

No configuration change needed.

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #4484

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- ~Documentation added/updated for [www.home-assistant.io][docs-repository]~ No configuration change needed.

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

Misiu commented:
Take a look at https://github.com/home-assistant/frontend/pull/7195.
You should probably cast `last_used_at` as Date.
This was suggested by @bramkragten in my PR.

aav7fl commented:
> `last_used_at`

Thanks! Great catch :)

bramkragten commented on src/panels/profile/ha-refresh-tokens-card.ts:
```diff
@@ -37,13 +37,25 @@ class HaRefreshTokens extends LitElement {
 
   protected render(): TemplateResult {
     const refreshTokens = this._refreshTokens(this.refreshTokens!);
+    function compareTokenLastUsedAt(tokenA, tokenB) {
```
If you want to create a method, move it our of the render function. Also, use a arrow function: `compareTokenLastUsedAt = (tokenA, tokenB) => {...}`

aav7fl commented:
@bramkragten Should be good to go now. 

- Moved the function out of the render
- Converted it to an arrow function
- Added strong typing so the comparator doesn't get used accidentally by something else

bramkragten commented on src/panels/profile/ha-refresh-tokens-card.ts:
```diff
@@ -35,6 +35,18 @@ class HaRefreshTokens extends LitElement {
       refreshTokens?.filter((token) => token.type === "normal").reverse()
   );
 
+  compareTokenLastUsedAt = (tokenA: RefreshToken, tokenB: RefreshToken) => {
```
you can move is outside of the class, as it doesn't use any class properties.
```suggestion
  const compareTokenLastUsedAt = (tokenA: RefreshToken, tokenB: RefreshToken) => {
```

bramkragten commented on src/panels/profile/ha-refresh-tokens-card.ts:
```diff
@@ -35,6 +35,18 @@ class HaRefreshTokens extends LitElement {
+  compareTokenLastUsedAt = (tokenA: RefreshToken, tokenB: RefreshToken) => {
+      const timeA = new Date(tokenA.last_used_at);
+      const timeB =  new Date(tokenB.last_used_at);
+      let comparison = 0;
+      if (timeA < timeB) {
+        comparison = 1;
+      } else if (timeA > timeB) {
+        comparison = -1;
+      }
+      return comparison;
```
We should return as soon as possible
```suggestion
      if (timeA < timeB) {
        return 1;
      } else if (timeA > timeB) {
        return -1;
      }
      return 0;
```

aav7fl replied:
To solve the `lint` issue, would I need to change the `new Date()` overload, or break it out and do `undefined` checks before comparing with the values?

bramkragten replied:
yes, you should check for undefined first, as:
```
new Date(undefined) -> Invalid Date
```

aav7fl replied:
Ok. I made a change in my last commit that should resolve that, but it looks like you might be suggesting a syntax I'm not familiar with.

bramkragten commented on src/panels/profile/ha-refresh-tokens-card.ts:
```diff
@@ -24,6 +24,21 @@ import {
 import { haStyle } from "../../resources/styles";
 import { HomeAssistant } from "../../types";
 
+const compareTokenLastUsedAt = (tokenA: RefreshToken, tokenB: RefreshToken) => {
+  const timeA = tokenA.last_used_at;
+  const timeB =  tokenB.last_used_at;
+  if (timeA === undefined) {
+    return 1;
+  } else if (timeB === undefined) {
+    return -1;
+  } else if (new Date(timeA) < new Date(timeB)) {
+    return 1;
+  } else if (timeA > timeB) {
+    return -1;
+  }
+  return 0;
+}
```
```suggestion
const compareTokenLastUsedAt = (tokenA: RefreshToken, tokenB: RefreshToken) => {
  const timeA = tokenA.last_used_at ? new Date(timeA) : 0;
  const timeB =  tokenB.last_used_at ? new Date(timeB) : 0;
  if (timeA < timeB) {
    return 1;
  } else if (timeA > timeB) {
    return -1;
  }
  return 0;
}
```

aav7fl replied:
oh duh🤦. I totally forgot about using a ternary operator when setting the const. Thanks!

bramkragten commented on src/panels/profile/ha-refresh-tokens-card.ts:
```diff
@@ -43,7 +55,7 @@ class HaRefreshTokens extends LitElement {
         ${this.hass.localize("ui.panel.profile.refresh_tokens.description")}
         ${refreshTokens?.length
-          ? refreshTokens!.map(
+          ? refreshTokens!.sort(compareTokenLastUsedAt).map(
```
We should do the sorting in the `memoizeOne`, to prevent doing it on every render, and only do it when the tokens are changed

aav7fl replied:
That makes sense. So change:

```ts
refreshTokens?.filter((token) => token.type === "normal").reverse()
```

to 

```ts
refreshTokens?.filter((token) => token.type === "normal").sort(compareTokenLastUsedAt)
```

?

zsarnett replied:
```ts
private _sortedTokens = memoizeOne((tokens) => tokens.sort(compareTokenLastUsedAt).map())
```

Something like this. Break it into a private function

bramkragten replied:
there is a `memoize` function already :-)

bramkragten replied:
@aav7fl yes, thats perfect!

aav7fl replied:
@bramkragten Change pushed. Thanks for walking through with me on this.

aav7fl committed:
Sort refresh tokens by 'last used at' (#4484)
